File: src/main/java/AoC2017_21.java
import static java.util.stream.Collectors.toList;

import java.util.Arrays;
import java.util.List;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public final class AoC2017_21 extends AoCBase {
    
    private static final CharGrid START = CharGrid.from(List.of(".#.", "..#", "###"));

    private final List<Rule> rules;
    
    private AoC2017_21(final List<String> inputs, final boolean debug) {
        super(debug);
        this.rules = inputs.stream().map(this::parseRule).collect(toList());
    }

    private Rule parseRule(final String input) {
        final String[] splits = input.split(" => ");
        return new Rule(
            CharGrid.from(Arrays.asList(splits[0].split("/"))),
            CharGrid.from(Arrays.asList(splits[1].split("/"))));
    }

    public static AoC2017_21 create(final List<String> input) {
        return new AoC2017_21(input, false);
    }

    public static AoC2017_21 createDebug(final List<String> input) {
        return new AoC2017_21(input, true);
    }
    
    private CharGrid enhance(final CharGrid grid) {
        return this.rules.stream()
            .filter(r -> r.from().contains(grid))
            .map(Rule::to)
            .findFirst().orElseThrow();
    }
    
    private long solve(final int iterations) {
        CharGrid grid = START;
        for (int i = 0; i < iterations; i++) {
            final int size = grid.getHeight();
            final CharGrid[][] subGrids = grid.divide(size % 2 == 0 ? 2 : 3);
            final CharGrid[][] enhanced = Arrays.stream(subGrids)
                .map(a ->
                    Arrays.stream(a).map(this::enhance).toArray(CharGrid[]::new))
                .toArray(CharGrid[][]::new);
            grid = CharGrid.merge(enhanced);
        }
        return grid.countAllEqualTo('#');
    }

    @Override
    public Long solvePart1() {
        return solve(5);
    }
    
    @Override
    public Long solvePart2() {
        return solve(18);
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2017_21.createDebug(TEST).solve(2) == 12L;

        final Puzzle puzzle = Aocd.puzzle(2017, 21);
        final List<String> input = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", create(input)::solvePart1),
            () -> lap("Part 2", create(input)::solvePart2)
        );
    }
    
    private static final List<String> TEST = splitLines(
        "../.# => ##./#../...\r\n" +
        ".#./..#/### => #..#/..../..../#..#"
    );
    
    record Rule(List<CharGrid> from, CharGrid to) {
        
        public Rule(final CharGrid from, final CharGrid to) {
            this(Utils.stream(from.getPermutations()).collect(toList()), to);
        }
    }
}

File: src/main/cpp/2017/21/AoC2017_21.cpp
#include <bits/stdc++.h>
#include "../../aoc/aoc.hpp"
#include "../../aoc/grid/grid.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

const Grid<char> START = Grid<char>::from({".#.", "..#", "###"});

struct Rule {
    Rule(const vector<Grid<char>>& from_, const Grid<char>& to_)
        : from(from_), to(to_) {};

    const vector<Grid<char>> from;
    const Grid<char> to;
};

Rule parseRule(const string& input) {
    const auto& splits = aoc::split(input, " => ");
    const Grid<char> from = Grid<char>::from(aoc::split(splits[0], "/"));
    const Grid<char> to = Grid<char>::from(aoc::split(splits[1], "/"));
    vector<Grid<char>> froms;
    transform(
            from.permutations_begin(), from.permutations_end(),
            back_inserter(froms),
            [](const Grid<char>& g) { return g; });
    return Rule(froms, to);
}

vector<Rule> parse(const vector<string>& input) {
    vector<Rule> rules;
    transform(input.begin(), input.end(),
            back_inserter(rules),
            [](const string& line) { return parseRule(line); });
    return rules;
}

Grid<char> enhance(const vector<Rule>& rules, const Grid<char>& grid) {
    for (const Rule& rule : rules) {
        if (find(rule.from.begin(), rule.from.end(), grid) == rule.from.end()) {
            continue;
        }
        return rule.to;
    }
    throw "No rule matched";
}

long solve(const vector<string>& input, const uint iterations) {
    const vector<Rule>& rules = parse(input);
    Grid<char> grid = START;
    for (uint i = 0; i < iterations; i++) {
        const uint size = grid.height();
        const vector<vector<Grid<char>>>& subGrids
            = grid.divide(size % 2 == 0 ? 2 : 3);
        vector<vector<Grid<char>>> enhanced;
        for (const vector<Grid<char>>& v : subGrids) {
            vector<Grid<char>> row;
            for (const Grid<char>& g : v) {
                row.push_back(enhance(rules, g));
            }
            enhanced.push_back(row);
        }
        grid = Grid<char>::merge(enhanced);
    }
    return grid.countAllEqualTo('#');
}

long part1(const vector<string>& input) {
    return solve(input, 5);
}

long part2(const vector<string>& input) {
    return solve(input, 18);
}

const vector<string> TEST = {
"../.# => ##./#../...",
".#./..#/### => #..#/..../..../#..#"
};

void samples() {
    assert(solve(TEST, 2) == 12);
}

MAIN(2017, 21)
